aws: Implementation of IAM Roles anywhere support in aws request signing (https://docs.aws.amazon.com/rolesanywhere/latest/userguide/introduction.html)

nbaws opened the pull request:
Commit Message: aws: Implementation of IAM Roles anywhere support in aws request signing
Additional Description: Creates an additional x509 credential type and adds custom signing support for IAM Roles Anywhere
Risk Level: Low
Testing: Unit
Docs Changes: doc update and sample config included
Release Notes: changelog update
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

nbaws commented:
@suniltheta if you have bandwidth

suniltheta reviewed (commented):
Couldn't go through all the files. Would it be possible to break the PR into multiple smaller chunks?

There are several functions, for example functions inside source/extensions/common/aws/signer_base_impl.cc or pemToAlgorithmSerialExpiration  which are tricky to review since I don't have context on how it should be written. If you are able to break down the PR into individual pieces you can link in code comments from the docs what that piece of function is trying to solve based on the guidance given in https://docs.aws.amazon.com/rolesanywhere/latest/userguide/authentication-sign-process.html or other docs page.

I see we have two new high level CredentailsProvider class IAMRolesAnywhereX509CredentialsProvider & IAMRolesAnywhereCredentialsProvider. Would it be possible to add docs comments to explain what each of these classes are doing? If you happen to have a short doc, will help with the getting the context.

nbaws commented:
> Couldn't go through all the files. Would it be possible to break the PR into multiple smaller chunks?
> 
> There are several functions, for example functions inside source/extensions/common/aws/signer_base_impl.cc or pemToAlgorithmSerialExpiration which are tricky to review since I don't have context on how it should be written. If you are able to break down the PR into individual pieces you can link in code comments from the docs what that piece of function is trying to solve based on the guidance given in https://docs.aws.amazon.com/rolesanywhere/latest/userguide/authentication-sign-process.html or other docs page.
> 
> I see we have two new high level CredentailsProvider class IAMRolesAnywhereX509CredentialsProvider & IAMRolesAnywhereCredentialsProvider. Would it be possible to add docs comments to explain what each of these classes are doing? If you happen to have a short doc, will help with the getting the context.

TY for your look through. I will add additional comments, and also carve off this code into its own source file so it is easier to understand. I will also add design documentation as per your suggestion.

nbaws commented on api/envoy/extensions/common/aws/v3/credential_provider.proto:
```diff
@@ -52,3 +59,29 @@ message AssumeRoleWithWebIdentityCredentialProvider {
+
+  string role_session_name = 7;
+
+  google.protobuf.Duration session_duration = 8 [(validate.rules).duration = {
```
If not specified there will be no duration sent, and the service will choose the duration. From https://docs.aws.amazon.com/rolesanywhere/latest/userguide/authentication-create-session.html:

durationSeconds
    The duration, in seconds, of the role session. The value is optional and can range from 900 seconds (15 minutes) up to 43200 seconds (12 hours).

nbaws commented:
@suniltheta I've refactored so this has less impact on existing source files. Fundamental code is the same and I've added a small design doc here: https://github.com/envoyproxy/envoy/issues/37440
Inline documentation can also be found in source/extensions/common/aws/iam_roles_anywhere_credentials_provider_impl.h

Please let me know if you'd like any more detail here.

nbaws commented on docs/root/configuration/http/http_filters/_include/aws_credentials.rst:
```diff
@@ -46,6 +46,10 @@ secret access key (the session token is optional).
    not set so that subsequently if the reloadable feature is set to ``true`` the cluster config is available to fetch the credentials.
 
+5. `IAM Roles Anywhere <https://docs.aws.amazon.com/rolesanywhere/latest/userguide/introduction.html>`_ can be used to retrieve IAM temporary credentials in exchange for an X509 Certificate.
+   A PEM formatted certificate, private key and certificate chain can be provided using the :ref:`credential_provider <envoy_v3_api_field_extensions.filters.http.aws_request_signing.v3.AwsRequestSigning.credential_provider>` field.
```
have added clarification to make PKCS#8 explicit

nbaws commented:
/retest

nbaws commented on source/extensions/common/aws/credentials_provider_impl.h:
```diff
@@ -409,6 +412,22 @@ class CredentialsProviderChainFactories {
+/**
+ * Credential provider based on an inline credential.
+ */
+class InlineCredentialProvider : public CredentialsProvider {
```
reverted changes to this entire file

nbaws commented on source/extensions/common/aws/credentials_provider_impl.cc:
```diff
@@ -1088,6 +1101,37 @@ absl::StatusOr<CredentialsProviderSharedPtr> createCredentialsProviderFromConfig
+    const std::string iam_roles_anywhere_endpoint = Utility::getRolesAnywhereEndpoint(region);
+    std::string role_session_name;
+    if (roles_anywhere.role_session_name().empty()) {
+      role_session_name = roles_anywhere.role_session_name();
```
ty for this pickup. adding test case to cover

nbaws commented on source/extensions/common/aws/utility.cc:
```diff
@@ -282,6 +282,10 @@ Utility::joinCanonicalHeaderNames(const std::map<std::string, std::string>& cano
 }
 
+std::string Utility::getRolesAnywhereEndpoint(absl::string_view region) {
+  return fmt::format("rolesanywhere.{}.amazonaws.com:443", region);
```
Fixed
